python/tank/deploy/git_descriptor.py

```diff
@@ -259,8 +188,95 @@ def _find_latest_by_pattern(self, pattern):
             new_loc_dict,
             self._type
         )
-        
 
+    def _find_latest_tag_by_pattern(self, version_numbers, pattern):
+        """
+        Given a list of version numbers, find the one that best matches the given pattern.
+
+        Version numbers passed in that doesn't match the pattern v1.2.3... will be ignored.
+
+        :param version_numbers: List of version number strings
+        :param pattern: Version patterns are on the following forms:
+
+            - v1.2.3 (can return this v1.2.3 but also any forked version under, eg. v1.2.3.2)
+            - v1.2.x (examples: v1.2.4, or a forked version v1.2.4.2)
+            - v1.x.x (examples: v1.3.2, a forked version v1.3.2.2)
+            - v1.2.3.x (will always return a forked version, eg. v1.2.3.2)
+
+        :returns: The most appropriate tag in the given list of tags
+        :raises: TankError if parsing fails
+        """
+        # now put all version number strings which match the form
+        # vX.Y.Z(.*) into a nested dictionary where it is keyed recursively
+        # by each digit (ie. major, minor, increment, then any additional
+        # digit optionally used by forked versions)
+        #
+        versions = {}
+        for version_num in version_numbers:
+            try:
+                version_split = map(int, version_num[1:].split("."))
+            except Exception, e:
+                # this git tag is not on the expected form vX.Y.Z where X Y and Z are ints. skip.
+                continue
+
+            if len(version_split) < 3:
+                # git tag has no minor or increment number. skip.
+                continue
+
+            # fill our versions dictionary
+            #
+            # For example, the following versions:
+            # v1.2.1, v1.2.2, v1.2.3.1, v1.4.3, v1.4.2.1, v1.4.2.2, v1.4.1,
+            #
+            # Would generate the following:
+            # {1: {2: {1: {}, 2: {}, 3: {1: {}}}, 4: {1: {}, 2: {1: {}, 2: {}}, 3: {}}}}
+            #
+            current = versions
+            for number in version_split:
+                if number not in current:
+                    current[number] = {}
+                current = current[number]
+
+        # now search for the latest version matching our pattern
+        version_to_use = None
+        if re.match("^v([0-9]+|x)(.([0-9]+|x)){2,}$", pattern):
+            # split our pattern, beware each part is a string (even integers)
+            version_split = re.findall("([0-9]+|x)", pattern)
+            if 'x' in version_split:
+                # check that we don't have an incorrect pattern using x
+                # then a digit, eg. v4.x.2
+                first_x = version_split.index('x')
+                if version_split[first_x:].count('x') < len(version_split)-first_x:
```

Comment:
Perhaps a comment here to explain what this is doing. Or maybe this could just be done with a regex to match any instance of an `x` followed by an integer?